src/coreclr/jit/lowerxarch.cpp

```diff
@@ -3693,6 +3720,70 @@ void Lowering::LowerHWIntrinsicToScalar(GenTreeHWIntrinsic* node)
         LowerNode(cast);
     }
 }
+
+//----------------------------------------------------------------------------------------------
+// Lowering::LowerAndOpToResetLowestSetBit: Lowers a tree AND(X, ADD(X, -1) to HWIntrinsic::ResetLowestSetBit
+//
+// Arguments:
+//    andNode - GT_AND node of integral type
+//
+// Return Value:
+//    Returns the replacement node if one is created else nullptr indicating no replacement
+//
+GenTree* Lowering::LowerAndOpToResetLowestSetBit(GenTreeOp* andNode)
+{
+    assert(andNode->OperIs(GT_AND) && varTypeIsIntegral(andNode));
+
+    GenTree* op1 = andNode->gtGetOp1();
+    GenTree* op2 = andNode->gtGetOp2();
+    if (op1->OperIs(GT_LCL_VAR) && op2->OperIs(GT_ADD) && !comp->lvaGetDesc(op1->AsLclVar())->IsAddressExposed())
+    {
+        GenTree* addOp1 = op2->gtGetOp1();
+        GenTree* addOp2 = op2->gtGetOp2();
+        if (addOp2->IsIntegralConst(-1) && addOp1->OperIs(GT_LCL_VAR) &&
+            (addOp1->AsLclVar()->GetLclNum() == op1->AsLclVar()->GetLclNum()) &&
+            ((op1->TypeIs(TYP_LONG) && comp->compOpportunisticallyDependsOn(InstructionSet_BMI1_X64)) ||
+             comp->compOpportunisticallyDependsOn(InstructionSet_BMI1)))
+        {
+            LIR::Use use;
+            if (BlockRange().TryGetUse(andNode, &use))
+            {
+                // If we allow the use of BSLR where the parent is NE or EQ it will generate "blsr, test"
+                // if we prevent the use of BSLR here then "OptimizeConstCompare" may generate TEST_NE or
```

Comment:
I don't get this part, C++ seems to be emitting blsr (a typo in the comment btw) in all cases: https://godbolt.org/z/fo8qGTe1q

I guess you mean the fact JIT doesn't realize blsr sets flags?

Reply:
When I showed the results to @SingleAccretion he suggested adding the code to check the use so that it would avoid the pattern. He'll have to explain the reasoning behind that.

Reply:
> I guess you mean the fact JIT doesn't realize blsr sets flags?

Yes. I suppose fixing it would require recognizing `blsr` here:

https://github.com/dotnet/runtime/blob/5fdc58d40141cc48d614a3e930f83f12cd538dad/src/coreclr/jit/lower.cpp#L2698-L2714

Reply:
I think that would just be adding the check `|| op1.OperIs(GT_HWINSTRINSIC) && (op1->AsHWIntrinsic()->GetHWIntrinsicId() == NamedIntrinsic::NI_BMI1_ResetLowestSetBit || op1->AsHWIntrinsic()->GetHWIntrinsicId() == NamedIntrinsic::NI_BMI1_X64_ResetLowestSetBit)`. Are you asking me to do this and remove the use check in the method I've added or is this just discussion?

Reply:
> Are you asking me to do this and remove the use check in the method I've added

I think it would be a good idea to do it as part of this change, yes.

Reply:
Done.
I had to change the instruction definition to include the flags so that they could later be checked. The diff is now all improvements, no regressions.